go-proio/reader.go

```diff
@@ -159,60 +174,98 @@ func (rdr *Reader) Get() (*Event, error) {
 	}
 
 	rdr.Err = err
+
+	rdr.getMutex.Unlock()
 	return event, err
 }
 
-//Events returns a channel of type Event where all of the events in the stream
-//will be pushed.  For error checking in this iteration scheme, check
-//Reader.Err.
-func (rdr *Reader) Events() <-chan *Event {
-	events := make(chan *Event)
+//ScanEvents returns a buffered channel of type Event where all of the events
+//in the stream will be pushed.  The channel buffer size is defined by
+//Reader.EventScanBufferSize which defaults to 100.  The goroutine responsible
+//for fetching events will not break until there are no more events,
+//Reader.StopScan() is called, or Reader.Close() is called.  For error checking
+//in this iteration scheme, check Reader.Err.
+func (rdr *Reader) ScanEvents() <-chan *Event {
+	events := make(chan *Event, rdr.EventScanBufferSize)
+
+	scanLoop := true
 	go func() {
-		for event, _ := rdr.Get(); event != nil; event, _ = rdr.Get() {
+		defer close(events)
+		for scanLoop {
+			event, _ := rdr.Get()
+			if event == nil {
+				break
+			}
 			events <- event
 		}
-		close(events)
 	}()
+
+	stopThisScan := func() {
+		scanLoop = false
```

Comment:
that's a race, I believe.

the canonical way to stop such a goroutine would be to:

```go
events := make(chan *Event, rdr.EventScanBufferSize)
quit := make(chan int)

go func() {
    defer close(events)
    for {
        evt, _ := rdr.Get()
        if evt == nil { break }
        select {
        case events <- evt:
        case <-quit:
            return
        }
    }
}()

stopThisScan := func () { close(quit) }()
```

Reply:
You're right, there is a possible race condition.  I should do it this way instead.